packages/eslint-plugin/src/rules/strict-enums.ts

```diff
@@ -0,0 +1,806 @@
+      }
+
+      /**
+       * Allow function calls that have a base enum that match the function
+       * type, like the following:
+       *
+       * ```ts
+       * function useFruit(fruit: Fruit) {}
+       * useFruit(Fruit.Apple);
+       * ```
+       */
+      if (setHasAnyElementFromSet(argumentEnumTypes, parameterEnumTypes)) {
+        return false;
+      }
+
+      return true;
+    }
+
+    // ------------------
+    // AST node callbacks
+    // ------------------
+
+    return {
+      /** When something is assigned to a variable. */
+      AssignmentExpression(node): void {
+        const leftType = getTypeFromNode(node.left);
+        const rightType = getTypeFromNode(node.right);
+
+        if (isAssigningNonEnumValueToEnumVariable(leftType, rightType)) {
+          context.report({
+            node,
+            messageId: 'mismatchedAssignment',
+          });
+        }
+      },
+
+      /** When a comparison between two things happen. */
+      BinaryExpression(node): void {
+        const leftType = getTypeFromNode(node.left);
+        const rightType = getTypeFromNode(node.right);
+
+        if (isMismatchedEnumComparison(node.operator, leftType, rightType)) {
+          context.report({
+            node,
+            messageId: 'mismatchedComparison',
+          });
+        }
+      },
+
+      /** When a function is invoked. */
+      CallExpression(node): void {
+        checkCallExpression(node);
+      },
+
+      /** When something is instantiated with the "new" keyword. */
+      NewExpression(node): void {
+        /**
+         * We need to perform the exact same checks on a class constructor
+         * invocation as we do on a normal function invocation.
+         */
+        checkCallExpression(node);
+      },
+
+      /** When a unary operator is invoked. */
+      UpdateExpression(node): void {
+        const argumentType = getTypeFromNode(node.argument);
+
+        /**
+         * Disallow using enums with unary operators, like the following:
+         *
+         * ```ts
+         * const fruit = Fruit.Apple;
+         * fruit++;
+         * ```
+         */
+        if (hasEnumTypes(argumentType)) {
+          context.report({
+            node,
+            messageId: 'incorrectIncrement',
+          });
+        }
+      },
+
+      /** When a new variable is created. */
+      VariableDeclaration(node): void {
+        for (const declaration of node.declarations) {
+          const leftTSNode =
+            parserServices.esTreeNodeToTSNodeMap.get(declaration);
+
+          /**
+           * Allow enum declarations without an initializer, like the following:
+           *
+           * ```ts
+           * let fruit: Fruit;
+           * if (something()) {
+           *   fruit = Fruit.Apple;
+           * } else {
+           *   fruit = Fruit.Banana;
+           * }
+           * ```
+           */
+          if (leftTSNode.initializer === undefined) {
+            continue;
+          }
```

Comment:
nitpick - avoid fetching the TS node till we really need it.
you can do this check purely with the ESTree AST:

```ts
if (declaration.init == null) {
  continue;
}

const rightType = getTypeFromNode(declaration.init);
```

----

Another thing we can do is ignore cases like this:
```ts
const x = Foo.A;
```
Why can we ignore this case? Because TS will always infer the type of `x` here - which means it'll match the initialised value.

So we can do a check like:
```ts
if (declaration.typeAnnotation == null) {
  continue;
}
```

to short-circuit, saving us time!

for context: TS types are computed lazily - so the less we fetch, the less work we do and thus the faster the rule is!

Reply:
The typeAnnotation property doesn't exist on either the node or the tsNode, so I'm not sure what you are referring to.